### PR TITLE
[REVIEW] Fixed core dump issue while passing dict_dtypes when `names=None` in read_csv()

### LINKED ISSUE #1198: [BUG] Passing a dict of dtypes to cudf.read_csv causes core dump
[Pandas's read_csv method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) accepts a dict of column names and dtypes:
```
dtype : Type name or dict of column -> type, optional
Data type for data or columns. E.g. {‘a’: np.float64, ‘b’: np.int32, ‘c’: ‘Int64’} Use str or object together with suitable na_values settings to preserve and not interpret dtype. If converters are specified, they will be applied INSTEAD of dtype conversion.
```
cudf.read_csv accepts a list of strings that must be in the same order as column names.

Description of Request:
cudf.read_csv should detect and support dtypes as either a list or a dict.

If dtype was supplied as a list, throw a warning noting that lists of dtypes are deprecated and users should supply a dict instead.

### PR DESCRIPTION
**Pull request by AK-ayush**
This PR fixes #1198 after completion. This error was coming because the case, when `names` is None and `dtypes` is a dict in `cudf.read_csv()`, is not handled in csv.py.

cc @harrism  
I need to discuss my approach with someone from `libcudf` side.  In the current approach, I am passing the key, value of dict_dtype as a string to same string array `csv_reader.dtype`. And another solution, I can think of creating a separate string array(i.e `csv_reader.dict_dtype`) in `csv_read_arg` structure and passing key, value to this string array. If there is any better approach, please do let me know.  

I'll add the tests later after discussion.

### CODE REVIEW

**Comment by harrism** [comment]
Since this is `csv_reader` I'll assign @vuule to review. CC @mjsamoht

**Comment by mjsamoht** [comment]
Wouldn't it be easier and cleaner to handle this in the Python layer and convert to a regular array before passing to `read_csv` in C++ layer?

**Comment by AK-ayush** [comment]
> Wouldn't it be easier and cleaner to handle this in the Python layer and convert to a regular array before passing to `read_csv` in C++ layer?

Yes but only if you know the column names before hand. So column names are not being passed then it is impossible to convert dict to regular array.

File: cpp/src/io/csv/csv_reader.cu [hunk 1]
```diff
@@ -750,7 +750,20 @@ gdf_error read_csv(csv_read_arg *args)
 		for ( int x = 0; x < raw_csv->num_actual_cols; x++) {
 
 			std::string temp_type 	= args->dtype[x];
-			gdf_dtype col_dtype		= convertStringToDtype( temp_type );
+			gdf_dtype col_dtype;
+			if(temp_type.find(':') != std::string::npos){
+			    for (auto it = raw_csv->col_names.begin(); it != raw_csv->col_names.end(); it++){
+			        std::size_t idx = temp_type.find(':');
+                    if(temp_type.substr( 0, idx) == *it){
```
Comment (harrism) [inline cpp/src/io/csv/csv_reader.cu]: Please correct the indentation.  Use whatever whitespace is being used in the rest of the file (tabs or spaces), please don't mix.

Comment (AK-ayush) [reply cpp/src/io/csv/csv_reader.cu]: I have fixed it. thanks

**Pull request merged**